server.go

```diff
@@ -15,13 +47,101 @@ type Server struct {
 
 	// Handler to invoke. If nil, uses DefaultHandler.
 	Handler Handler
+
+	// Shutdown handling
+	// --
+	lock         sync.RWMutex
+	started      bool
+	wg           sync.WaitGroup            // active workers (servers)
+	tcpListeners map[net.Listener]struct{} // track active TCP listeners
+	udpListeners map[*net.UDPConn]struct{} // track active UDP listeners
+	tcpConns     map[net.Conn]struct{}     // track active connections
+}
+
+// type check
+var _ ServerDNSCrypt = &Server{}
+
+// Shutdown - tries to gracefully shutdown the server. It waits until all
+// connections are processed and only after that it leaves the method.
+// If context deadline is specified, it will exit earlier.
+func (s *Server) Shutdown(ctx context.Context) error {
+	log.Info("Shutting down the DNSCrypt server")
+
+	s.lock.Lock()
+	if !s.started {
+		s.lock.Unlock()
+		log.Info("Server is not started")
+		return ErrServerNotStarted
+	}
+
+	s.started = false
+
+	// Unblock reads for all active tcpConns
+	for conn := range s.tcpConns {
+		_ = conn.SetReadDeadline(time.Unix(1, 0))
+	}
+
+	// Unblock reads for all active TCP listeners
+	for l := range s.tcpListeners {
+		switch v := l.(type) {
+		case *net.TCPListener:
+			_ = v.SetDeadline(time.Unix(1, 0))
```

Comment:
1.  The value of `time.Unix(1, 0)` is used several times.  Maybe just assign it to a variable once?
2.  All those `SetFoo` really need a comment explaining why it is done that way instead of a simple `Close()`.